TestCases/compliance-level-3/0068-feel-equality/0068-feel-equality.dmn

```diff
@@ -0,0 +1,727 @@
+        <allowedAnswers>false (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="deep_003"/>
+        <literalExpression>
+            <text>[1,2,[3, 4]] = [1,2,[4, 3]]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="deep_004" id="_deep_004">
+        <description>Tests FEEL expression: '[1,2,{a: [3,4]}] = [1,2,{a: [3,4], b: "foo"}]' and expects result: 'false (boolean)'</description>
+        <question>Result of FEEL expression '[1,2,{a: [3,4]}] = [1,2,{a: [3,4], b: "foo"}]'?</question>
+        <allowedAnswers>false (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="deep_004"/>
+        <literalExpression>
+            <text>[1,2,{a: [3,4]}] = [1,2,{a: [3,4], b: "foo"}]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="deep_005" id="_deep_005">
+        <description>Tests FEEL expression: '{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [1,2]}}' and expects result: 'true (boolean)'</description>
+        <question>Result of FEEL expression '{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [1,2]}}'?</question>
+        <allowedAnswers>true (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="deep_005"/>
+        <literalExpression>
+            <text>{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [1,2]}}</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="deep_006" id="_deep_006">
+        <description>Tests FEEL expression: '{a: {c: [1,2], b: "foo"}} = {a: {b: "foo", c: [1,2]}}' and expects result: 'true (boolean)'</description>
+        <question>Result of FEEL expression '{a: {c: [1,2], b: "foo"}} = {a: {b: "foo", c: [1,2]}}'?</question>
+        <allowedAnswers>true (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="deep_006"/>
+        <literalExpression>
+            <text>{a: {c: "bar", b: "foo"}} = {a: {b: "foo", c: "bar"}}</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="deep_007" id="_deep_007">
+        <description>Tests FEEL expression: '{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [2,1]}}' and expects result: 'false (boolean)'</description>
+        <question>Result of FEEL expression '{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [2,1]}}'?</question>
+        <allowedAnswers>false (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="deep_007"/>
+        <literalExpression>
+            <text>{a: {b: "foo", c: [1,2]}} = {a: {b: "foo", c: [2,1]}}</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="singleton_001" id="_singleton_001">
+        <description>Tests FEEL expression: '[100] = 100' and expects result: 'false (boolean)'</description>
+        <question>Result of FEEL expression '[100] = 100'?</question>
+        <allowedAnswers>false (boolean)</allowedAnswers>
+        <variable typeRef="boolean" name="singleton_001"/>
+        <literalExpression>
+            <text>[100] = 100</text>
```

Comment:
type(`[100]`) is FEEL list
type(`100`) is FEEL number
This is another instance of a test for equality of two different types, and the test case expect `false` but again the spec
![image](https://user-images.githubusercontent.com/1699252/51900752-5b0b5480-23b6-11e9-934d-5b000b6e6f0a.png)

sorry :)

Reply:
I know this is your intention, but still the test as-is I believe is conflicting with the spec snippet I pasted above, and further semantic of equality from this snippet too:
![image](https://user-images.githubusercontent.com/1699252/51902030-57c59800-23b9-11e9-91ee-365057ed9642.png)
So I believe we need some other way to test singleton non-equality, which must be _also_ in the boundary of the spec ;) I am thinking on it but with no solution at the moment from my side

Reply:
I'm not sure it matters here whether they are different types or not.  In a sense the test asserts two things - singleton non-equality and different type non-equality.  However, without an explicit test like this I am not sure how we would assert singleton non-equality.  We have to assert it _somehow_ and without a test like vendors may not know they are not 1.2 compliant. 

I'm not so sure that tests need to be within the boundaries of the spec - by their very nature they test the boundaries of the spec.  Like a test for asserting too many params passed to a function, or the wrong typed params - it is the same thing in my view.

Reply:
By the spec snippet(s) I pasted above, if I interpret it correctly and I would be happy to be corrected, an expression like `[100] = 100` could be either `false` or `null` (and an error).
Your test is enforcing for only the `false` option instead.
That is btw one of the reason we avoided in the TCK to test for errors, to avoid this kind of cases. 

I am not saying I don't want to test singleton list equality; I am saying I would prefer for a test which doesn't enforce for a single outcome when the spec (in my view) is allowing both.

Reply:
How about (but I would need to verify it) testing for:
```
([100] = 100) != true
```
and expecting `true`?
Again, just wild guessing and I would need later to verify. Am I better able with this pragmatical example to demonstrate better what I am trying to convey?

Reply:
Looking at the equality section in the spec I can't see anything that says an equality check between different types is null. The check has to include type equivalence but I think it should just return false not null when types are not equal. 

Or have you seen something else in the spec?  Heaven knows stuff can be a bit spread out.

Reply:
> Or have you seen something else in the spec?

![image](https://user-images.githubusercontent.com/1699252/51904013-d91f2980-23bd-11e9-890f-c406fafc203e.png)

Reply:
Thanks, but oh dear. That sucks.

Reply:
LOL. My suggestion would be to hold (i.e.: comment out) this test and I will raise a specific issue to the RTF to ask for clarification, pointing out these 2 snippet(s) from the spec. Wdyt ?

Reply:
The trouble is, it means that `123 = "123"` is null, and likely `123 = null` is null also.  I really can't imagine why the spec designers would consider such a thing.  A violation of "the principle of least surprise".  

It also means a test like `([100] = 100) != true` would be comparing null to true, which would evaluate to null ... so it may as well be `([100] = 100) != "hamburger"`.

However, note your above spec snippet reference begins with 'In general' which is not exactly concrete, so I suspect this may be a reference to the date time thing: 

> "Where necessary, including the value dt function (see 10.3.2.3.6), a date value is considered to be equivalent to a date time value in which the time of day is UTC midnight (00:00:00)."

So actually, I think the tests should be changed to assert that `[100] = 100` is null, as well as `123 = "123"` is null - as *silly* as it seems IMHO.  But, I think you've shown the spec says it as such. 

But .. we should raise a concern with the RTF that the equality thing is nuts and ill-defined.  And seek clarification on what 'In general' means, as well as what "where necessary" means for date and datetime equality. 

I do think we need to keep the `[100] = 100` test in some guise to assert the non-equality of singletons - regardless of whether it is null or false.  But I think, as you've shown, it should be null.

Reply:
I believe I just found another section of the DMNv1.2 spec where it contradicts that `[a]=a` is false, and expecting to be true. I am going to raise a specific issue and I will keep posted.

Reply:
Ref: https://issues.omg.org/browse/INBOX-808

The spec says:
![image](https://user-images.githubusercontent.com/1699252/51988921-31c9f180-24a6-11e9-8153-8342e9c94fe2.png)
but
```
get entries({key1 : "value1"})[key="key1"].value = "value1"

  by substitution:

[ { key : "key1", value : "value1" } ][key="key1"].value = "value1"
[ { key : "key1", value : "value1" } ].value = "value1"
[ "value1" ] = "value1"
```

and the expecting to be true. Hence the DMNv1.2 spec seems to be contradicting this proposed test.

In conclusion, if this will be the only "controversial" test I would eventually propose to comment out this specific case only, pending for the linked OMG RTF ticket to be addressed ;)